experiments/experiments.go

```diff
@@ -0,0 +1,273 @@
+	// experiment means all variant calls will return an empty string.
+	enabled bool
+	// startTime determines when this experiment is due to start. Variant
+	// requests prior to this time will return an empty string.
+	startTime time.Time
+	// endTime determines when this experiment is due to end. Variant requests
+	// after this time will return an empty string.
+	endTime time.Time
+	// logBucketing determines whether bucketing events should be logged.
+	logBucketing bool
+	// bucketVal is a string used for shifting the deterministic bucketing
+	// algorithm.  In most cases, this will be an Account's fullname.
+	bucketVal string
+	// variantSet contains a set of experimental variants as well as their
+	// distributions. It is used by experiments to track which bucket a variant
+	// is assigned to.
+	variantSet VariantSet
+}
+
+// NewSimpleExperiment returns a new instance of SimpleExperiment. Default
+// values if not otherwise provided by the ExperimentConfig will be assumed.
+func NewSimpleExperiment(experiment *ExperimentConfig) (*SimpleExperiment, error) {
+	shuffleVersion := experiment.Experiment.ShuffleVersion
+	if shuffleVersion == "" {
+		shuffleVersion = "None"
+	}
+	bucketVal := experiment.Experiment.BucketVal
+	if bucketVal == "" {
+		bucketVal = "user_id"
+	}
+	enabled := true
+	if experiment.Enabled != nil {
+		enabled = *experiment.Enabled
+	}
+	bucketSeed := experiment.Experiment.BucketSeed
+	if experiment.Experiment.BucketSeed == "" {
+		bucketSeed = fmt.Sprintf("%d.%s.%s", experiment.ID, experiment.Name, shuffleVersion)
+	}
+	variantSet, err := FromExperimentType(experiment.Type, experiment.Experiment.Variants, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &SimpleExperiment{
+		id:         experiment.ID,
+		name:       experiment.Name,
+		bucketSeed: bucketSeed,
+		bucketVal:  bucketVal,
+		enabled:    enabled,
+		startTime:  time.Unix(experiment.StartTimestamp, 0),
+		endTime:    time.Unix(experiment.StopTimestamp, 0),
+		numBuckets: 1000,
+		variantSet: variantSet,
+	}, nil
+}
+
+// Variant determines the variant, if any, is active. Bucket calculation is
+// determined based on the bucketVal.
+func (e *SimpleExperiment) Variant(args map[string]string) (string, error) {
+	if !e.isEnabled() {
+		return "", nil
+	}
+	args = lowerArguments(args)
+	if value, ok := args[e.bucketVal]; !ok || value == "" {
+		return "", fmt.Errorf("must specify %s in call to variant for experiment %s", e.bucketVal, e.name)
+	}
+	bucket := e.calculateBucket(args[e.bucketVal])
+	return e.variantSet.ChooseVariant(bucket), nil
+}
+
+func lowerArguments(args map[string]string) map[string]string {
+	lowered := make(map[string]string, len(args))
+	for key, value := range args {
+		lowered[strings.ToLower(key)] = value
+	}
+	return lowered
+}
+
+func (e *SimpleExperiment) calculateBucket(bucketKey string) int {
+	target := new(big.Int)
+	bucket := new(big.Int)
+	hashed := sha1.Sum([]byte(fmt.Sprintf("%s%s", e.bucketSeed, bucketKey)))
```

Comment:
nit: The performance of `fmt.Sprintf` is not very good (because the unknown of the length of the final result and the allocations happened), for this case `e.bucketSed + bucketKey` would be faster and you don't lose too much readability (for most of the other cases in this package `fmt.Sprintf` is still much more readable and convenient)